src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,241 @@
+
+    bool value_contained = std::binary_search(dictionary->cbegin(), dictionary->cend(), t_value);
+
+    if (!value_contained) {
+      if (scan_type == ScanType::OpEquals) {
+        // no value can be equal
+        continue;
+      }
+      if (scan_type == ScanType::OpNotEquals) {
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+    }
+
+    ValueID comp_value;
+    switch (scan_type) {
+      case ScanType::OpEquals:
+      case ScanType::OpNotEquals:
+      case ScanType::OpLessThan:
+      case ScanType::OpGreaterThanEquals:
+        comp_value = dict_column->lower_bound(t_value);
+        break;
+      case ScanType::OpLessThanEquals:
+      case ScanType::OpGreaterThan:
+        comp_value = dict_column->upper_bound(t_value);
+        break;
+    }
+
+    if (comp_value == ValueID{0u}) {
+      if (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpLessThan) {
+        // value is smaller than all the values in the column, there are no matches
+        // we can go to the next chunk
+        continue;
+      }
+      if (scan_type == ScanType::OpGreaterThan || scan_type == ScanType::OpGreaterThanEquals) {
+        // every value is greater than the search value - add every position
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+    }
+
+    // if upper_bound goes past the end
+    if (comp_value == INVALID_VALUE_ID) {
+      if (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpLessThan) {
+        // value is greater than all the values in the column
+        add_all_values(*pos_list, chunk_id, dict_column->size());
+        continue;
+      }
+      if (scan_type == ScanType::OpGreaterThan || scan_type == ScanType::OpGreaterThanEquals) {
+        // every value is smaller than the search value
+        continue;
+      }
+    }
+
+    if (comp_value != ValueID{0} && comp_value != INVALID_VALUE_ID &&
+        (scan_type == ScanType::OpLessThanEquals || scan_type == ScanType::OpGreaterThan)) {
+      comp_value--;
+    }
+
+    auto attribute_vector = dict_column->attribute_vector();
+
+    for (ChunkOffset i = 0; i < attribute_vector->size(); ++i) {
+      if (ValueID_comparator(attribute_vector->get(i), comp_value)) {
+        pos_list->push_back(RowID{chunk_id, i});
+      }
+    }
+  }  // end for (chunks)
+
+  Chunk reference_chunk;
+
+  // create reference columns
+  for (ColumnID i{0}; i < table->col_count(); ++i) {
+    auto reference_column = std::make_shared<ReferenceColumn>(table, i, pos_list);
+    reference_chunk.add_column(reference_column);
+  }
+
+  output_table->emplace_chunk(std::move(reference_chunk));
+
+  return output_table;
+}
+
+void add_all_values(PosList& pos_list, ChunkID chunk_id, ChunkOffset number_of_values) {
+  pos_list.reserve(number_of_values);
+  for (ChunkOffset i = 0; i < number_of_values; i++) {
+    pos_list.push_back(RowID{chunk_id, i});
```

Comment:
You could've also used `emplace_back`, now that you've already reserved the neccesary space.